std/algorithm/sorting.d

```diff
@@ -1048,147 +1048,294 @@ unittest
-    rs = The ranges to compute the union for.
-
-Returns:
-    A range containing the union of the given ranges.
-
-See_Also:
-   $(XREF algorithm, setops, SetUnion)
- */
-struct Merge(alias less = "a < b", Rs...) if (allSatisfy!(isInputRange, Rs))
+struct Merge(alias less = "a < b", Rs...)
+    if (Rs.length >= 2 &&
+        allSatisfy!(isInputRange, Rs) &&
+        !is(CommonType!(staticMap!(ElementType, Rs)) == void))
 {
-private:
-    Rs _r;
-    alias comp = binaryFun!(less);
-    uint _crt;
+    import std.functional : binaryFun;
+    import std.traits : isCopyable;
+    import std.typetuple : anySatisfy;
+
+    private alias comp = binaryFun!less;
+    private alias ElementType = CommonType!(staticMap!(.ElementType, Rs));
+    private enum isBidirectional = allSatisfy!(isBidirectionalRange, staticMap!(Unqual, Rs));
 
-    void adjustPosition(uint candidate = 0)()
+    this(Rs source)
     {
-        static if (candidate == Rs.length)
+        this.source = source;
+        this._lastFrontIndex = frontIndex;
+        static if (isBidirectional)
         {
-            _crt = _crt.max;
+            this._lastBackIndex = size_t.max; // uinitialized, lazy initialized in `front`
         }
-        else
+    }
+
+    static if (anySatisfy!(isInfinite, Rs))
+    {
+        enum bool empty = false; // propagate infiniteness
+    }
+    else
+    {
+        @property bool empty()
         {
-            if (_r[candidate].empty)
-            {
-                adjustPosition!(candidate + 1)();
-                return;
-            }
-            foreach (i, U; Rs[candidate + 1 .. $])
-            {
-                enum j = candidate + i + 1;
-                if (_r[j].empty) continue;
-                if (comp(_r[j].front, _r[candidate].front))
-                {
-                    // a new candidate was found
-                    adjustPosition!(j)();
-                    return;
-                }
-            }
-            // Found a successful candidate
-            _crt = candidate;
+            return _lastFrontIndex == size_t.max;
         }
     }
 
-public:
-    alias ElementType = CommonType!(staticMap!(.ElementType, Rs));
-    static assert(!is(CommonType!(staticMap!(.ElementType, Rs)) == void),
-        typeof(this).stringof ~ ": incompatible element types.");
-
-    ///
-    this(Rs rs)
+    @property auto ref front()
     {
-        this._r = rs;
-        adjustPosition();
+        final switch (_lastFrontIndex)
+        {
+            foreach (i, _; Rs)
+            {
+            case i:
+                assert(!source[i].empty);
+                return source[i].front;
+            }
+        }
     }
 
-    ///
-    @property bool empty()
+    private size_t frontIndex()
     {
-        return _crt == _crt.max;
+        size_t bestIndex = size_t.max; // indicate undefined
+        Unqual!ElementType bestElement;
+        foreach (i, _; Rs)
+        {
+            if (source[i].empty) { continue; }
```

Comment:
This is unusual, for single statements on one line we omit the braces (except when required e.g. for function definitions).